src/Avalonia.Controls/Control.cs

```diff
@@ -84,6 +84,13 @@ public class Control : InputElement, IControl, INamed, IVisualBrushInitialize, I
                 nameof(Unloaded),
                 RoutingStrategies.Direct);
 
+        /// <summary>
+        /// Defines the <see cref="SizeChanged"/> event.
+        /// </summary>
+        public static readonly RoutedEvent<SizeChangedEventArgs> SizeChangedEvent =
+            RoutedEvent.Register<Control, SizeChangedEventArgs>(
+                nameof(SizeChanged), RoutingStrategies.Direct);
```

Comment:
Both WPF and UWP use CLR events. Is there a point of using routed events? And yeah, it should be direct only if routed.

Reply:
Are you sure? WPF is a routed event and I can't check the WinUI source here.

https://github.com/dotnet/wpf/blob/bba0c14e636cb6fa3e6108e2d7a81b70cca5c139/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/FrameworkElement.cs#L3319-L3337

The EventArgs in both WPF/WinUI are deriving from RoutedEventArgs too so it would be a mistake if these are actually CLR events.

However, a CLR event would work just fine and have less overhead in this case. I can change it if needed but don't think it's really necessary and would somewhat break compatibility with WPF from what I see.

Reply:
UWP docs say its equivalent to Direct RoutedEvent:
> Although this event uses a [RoutedEventHandler](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.routedeventhandler?view=winrt-22621)-related delegate and a [RoutedEventArgs](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.routedeventargs?view=winrt-22621)-derived class as event data, the event is not truly a routed event. It doesn't bubble through an object tree. It can be handled only on the element that originates the event (in other words, the sender). [OriginalSource](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.routedeventargs.originalsource?view=winrt-22621#windows-ui-xaml-routedeventargs-originalsource) in event data for this event is always null, so don't try to use the [OriginalSource](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.routedeventargs.originalsource?view=winrt-22621#windows-ui-xaml-routedeventargs-originalsource).

Reply:
Yeah, seems I misread the docs page.

Reply:
It is buried way at the bottom of the page. I didn't get that far either and was basing this implementation on WPF anyway. WPF has more functionality here. However, this is a place they changed UWP/WinUI compared to WPF.

I would prefer to keep the WPF-based implementation if there are no objections (even though there is a theoretical performance benefit with the UWP one).